javascript/stimulus_reflex.js

```diff
@@ -85,12 +85,21 @@ const invokeLifecycleMethod = (stage, reflex, element) => {
 const createSubscription = controller => {
   const { channel, room } = controller.StimulusReflex
   const id = `${channel}${room}`
-
   const subscription =
     app.StimulusReflex.subscriptions[id] ||
     app.StimulusReflex.consumer.subscriptions.create(
       { channel, room },
       {
+        connected: () =>
+          document.body.classList.replace(
+            'reflex-disconnected',
+            'reflex-connected'
+          ),
+        disconnected: () =>
+          document.body.classList.replace(
+            'reflex-connected',
+            'reflex-disconnected'
+          ),
```

Comment:
Note that each SR controller gets it's own subscription, so we'll need to move the css class to the respective controller's element. It can be used in the same manner that you proposed for the body class.

Also, I think we can get away with a single class and simplify this further.

```javascript
connected: () => controller.element.classList.add('reflex-connected'),
disconnected: () => controller.element.classList.remove('reflex-connected'),
```

Reply:
I originally got confused on this, too. However, this code path only executes once because of the clever way you evaluate it.

I'm 100% confident that this is supposed to be on the body element, however you want to select it.

Reply:
FWIW, I like having classes for both. That way you don't have to test for the absence of, which is kind of ugly from a selector perspective.

Reply:
Ah yes you're correct about body. I still propose that we change this to be a single `connected` css class.

Reply:
I've explained why I proposed two separate classes. Could you explain why you want to push for one? You mentioned simplification but I don't think that's what this would accomplish. There is the `:not()` selector but `body.reflex-disconnected` is far easier to work with.

Essentially, I am suggesting that the option to craft cascades based on the potential for both classes has more flexibility than `:not()` and as a hack front-end dev I'm probably not the best avatar to defend the position, which sucks, because I think I'm right.

Reply:
My thinking is more along the lines of mobile first responsive design. Set the default styles to be non reflex connected. i.e. `hidden`, `disabled`, etc... Then add custom css for when the reflex is connected. This feels simpler to my mind.

Reply:
Essentially this philosophy applied to SR _(where disconnected is default)_.

> we present the mobile layout and functionality as the default configuration provided, before additional information is loaded on top of that

https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps/Responsive/Mobile_first

Reply:
What originally triggered my thinking on this was the use of `classList.replace` which will only work if the `disconnected` class has already been added to the `body` tag.

Reply:
This is true, which is technically why I did [this](https://github.com/pjforde1978/stimulus_reflex/blob/772e27df9d360d082a402f6b8c6a6e7d93348603/javascript/stimulus_reflex.js#L273) - **it starts off in the reflex-disconnected state**.

Here's how I set up my snake example to take advantage of the PR:

```Scss
body.reflex-connected .disconnected {
  display: none;
}
body.reflex-disconnected .connected {
  display: none;
}
```
```Erb
<h3 class="text-info disconnected">Connecting...</h3>
<button class="btn btn-primary connected">Start</button>
```

The idea, of course, is that you want to show "Connecting" when disconnected, and the "Start" button when connected. What's not to love?

To do it "your" way :grinning: looks like this:

```Scss
body.reflex-connected .disconnected {
  display: none;
}
body:not(.reflex-connected) .connected {
  display: none;
}
```

If you honestly find that more clear, then we simply disagree and I'll follow your lead without further grief because ultimately, you have to be the decider.

Reply:
Either way, it's worth celebrating that this will take an additional 9 LOC off the JS controller.

Reply:
If we invert my proposal above and make the default state `connected`, the CSS would change to this.

```css
.disconnected { display: none; }
body.reflex-disconnected .disconnected { display: auto; }
body.reflex-disconnected .connected { display: none; }
```

```html
<div class="disconnected">Connecting</div>
<div class="connected">Connected</div>
```

Reply:
Yeah, that’s acceptable. Not the hill I die on, either. I really will support whatever you run with.